File: src/evaluation/evaluate.py
# from ranking import kendallsTau
# from pathlib import Path
# from scipy.stats import kendalltau
# import os


# DATASETS_DIRS = ['../../data/multiplex/extracted/', '../../data/multilayer/generated']

# SIMULATIONS_DIR = '../../results/simulations/'
# ALGORITHMS_DIR = '../../results/algorithms/'
# RANKING_DIR = '../../results/rankings/'

# AVG_FILE = 'avg'

# # ALGORITHMS = ['addPageRank', 'verPageRank', 'laPCI', 'mlPCI', 'lsPCI', 'aggPageRank', 
# #         'multiCore', 'sumCore', 'aggCore', 'alPCI', 'aggDegree', 'verBetweenness']

# def read_alg_scores(net_path):
#     res = {}
#     for alg in os.listdir(net_path):
#         if alg.endswith('.alg'):
#             score = []
#             with open(net_path + '/' + alg) as f:
#                 t, _ = f.readline().split() # ignore header line
#                 type_f = int if t == 'i' else float
#                 for k in f:
#                     score.append(type_f(k))
#             res[alg.rstrip('.alg')] = score 
#     return res

# def compute_avg(path):
#     count = 0
#     score = None
#     for filename in os.listdir(path):
#         if filename.endswith('.sim'):
#             count += 1
#             with open(path + '/' + filename) as f:
#                 num_nodes, _ = map(int, f.readline().split())
#                 score = score or [0 for _ in range(num_nodes)]
#                 for i, k in enumerate(f):
#                     score[i] += int(k)

#     with open(path + '/' + AVG_FILE, 'w+') as f:
#         for s in score:
#             f.write(str(s / count) + "\n")

# def read_sim_scores(path):
#     score = []
#     with open(path + '/' + AVG_FILE) as f:
#         for r in f:
#             score.append(float(r))
#     return score

# def evaluate(sim_score, alg_scores):
#     return {alg_name : float(kendallsTau(alg_score, sim_score)) for alg_name, alg_score in alg_scores.items()}

# def write_rankings(net, lambda_dir, ktt):
#     Path(RANKING_DIR).mkdir(parents=True, exist_ok=True)
#     Path(RANKING_DIR + net).mkdir(parents=True, exist_ok=True)
#     with open(RANKING_DIR + net + '/' + lambda_dir, 'w+') as f:
#         for name, score in ktt.items():
#             f.write("%s %f\n" % (name, score))

# if __name__ == "__main__":
#     # for each dataset output the kendall's Tau
#     for dir in DATASETS_DIRS:
#         for net in os.listdir(dir):
#             net = net.rstrip('.edges')
#             # print("#%s" % net)
#             alg_scores = read_alg_scores(ALGORITHMS_DIR + net)
#             for lambda_dir in os.listdir(SIMULATIONS_DIR + net):
#                 sim_net_lambda_dir = SIMULATIONS_DIR + net + '/' + lambda_dir
#                 if not AVG_FILE in os.listdir(sim_net_lambda_dir):
#                     compute_avg(sim_net_lambda_dir)
#                 sim_scores = read_sim_scores(sim_net_lambda_dir)
#                 ktt = evaluate(sim_scores, alg_scores)
#                 write_rankings(net, lambda_dir, ktt)
#             # # dictionary with different spreading powers
#             # simulations_ranking = read_simulations_output(net)
#             # for algorithm in ALGORITHMS:
#             #     algorithm_ranking = read_algorithm_output(algorithm, net)
#             #     kt = {}
#             #     for sp in simulations_ranking:
#             #         kt[sp] = kendallsTau(simulations_ranking[sp], algorithm_ranking)
#             #     write_result(net, algorithm, kt)

File: src/evaluation/evaluate.cpp
#include <fstream>
#include <cassert>
#include <iostream>
#include "ranking.h"
#include "io.h"


string DATASETS_DIRS[2] = {"../../data/multiplex/extracted/", "../../data/multilayer/generated"};

string SIMULATIONS_DIR = "../../results/simulations/";
string ALGORITHMS_DIR = "../../results/algorithms/";
string RANKING_DIR = "../../results/rankings/";

string AVG_FILE = "avg";

// ALGORITHMS = ["addPageRank", "verPageRank", "laPCI", "mlPCI", "lsPCI", "aggPageRank", 
//         "multiCore", "sumCore", "aggCore", "alPCI", "aggDegree", "verBetweenness"]

bool endswith(string &value, string ending){
    return equal(ending.rbegin(), ending.rend(), value.rbegin());
}

void read_alg_scores(string &net_path, vector<string> &alg_names, vector<vector<double>> &alg_scores){
    vector<string> files;
    // cerr << net_path << endl;
    read_directory(net_path, files);
    // cerr << "alg dir read" << endl;
    for (string &alg : files){
        if (endswith(alg, ".alg")){
            cerr << "\t" << alg << endl; 
            ifstream in(net_path + "/" + alg);
            char t;
            int n;

            in >> t >> n;
            alg_names.push_back(alg.substr(0, alg.size() - 4));
            alg_scores.push_back(vector<double>(n, 0));
            
            vector<double> &score = alg_scores[alg_scores.size() - 1];
            // cerr << "scores initialized for " << alg << endl;
            for(int i = 0; i < n; i++)
                in >> score[i];
            // cerr<< "scores read" << endl;
        }
    }
}

void compute_avg(string &path){
    int count = 0;
    vector<int> score;
    vector<string> files;
    read_directory(path, files);
    for (string &filename : files){
        if (endswith(filename, ".sim")){
            count += 1;
            ifstream in(path + "/" + filename);
            int num_nodes, sum_nodes;
            in >> num_nodes >> sum_nodes;
            if(score.size() == 0)
                score.resize(num_nodes, 0);
            int k;
            for(int i = 0; i < num_nodes; i++){
                in >> k;
                score[i] += k;
            }
        }
    }
    ofstream out(path + "/" + AVG_FILE);

    out << score.size() << endl;
    for (int s : score)
        out << (double(s) / count) << endl;
}

vector<double> read_sim_scores(string &path){
    ifstream in(path + "/" + AVG_FILE);
    int n;
    in >> n;
    vector<double> score(n, 0);
    for(int i = 0; i < n; i++)
        in >> score[i];
    
    return score;
}

vector<double> evaluate(vector<double> &sim_score, vector<vector<double>> &alg_scores){
    vector<double> scores(alg_scores.size(), 0);
    int i = 0;
    for(vector<double> &s : alg_scores){
        // cout << "s:" << s.size() << ", " << "sim:" << sim_score.size() << endl;
        assert(s.size() == sim_score.size());
        scores[i] = kendallsTau(s, sim_score);
        i++;
    }   
    return scores;
}


void write_rankings(string &base, string &net, string &lambda_dir, vector<string> &alg_names, vector<double> &ktt){
    string rank_dir = base + "/" + RANKING_DIR;
    cerr << "# " << rank_dir << endl;
    if(!dir_exists(rank_dir)){
        create_dir(rank_dir);
    }
    string ranking_net_dir = rank_dir + net;

    if(!dir_exists(ranking_net_dir)){
        create_dir(ranking_net_dir);
    }

    ofstream out(ranking_net_dir + "/" + lambda_dir);
    cerr << (ranking_net_dir + "/" + lambda_dir) << endl;

    for(int i = 0; i < alg_names.size(); i++){
        cerr << "\t" << alg_names[i] << endl;
        out << alg_names[i] << " " << ktt[i] << endl;}
}

int main(int argc, char const *argv[]) {

    string argv_str(argv[0]);
    string base = argv_str.substr(0, argv_str.find_last_of("/"));

    // # for each dataset output the kendall"s Tau
    for (string &dir : DATASETS_DIRS){
        vector<string> nets;
        string dir_path = base + "/" + dir;
        read_directory(dir_path, nets);
        for (string &net : nets){
            string net_name;
            // cerr << ">" << net << endl;
            if(endswith(net, ".edges"))
                net_name = net.substr(0, net.size() - 6);
            else
                net_name = net;
            cerr << net_name << endl;
            string net_path = base + "/" + ALGORITHMS_DIR + net_name;
            vector<string> alg_names;
            vector<vector<double>> alg_scores;
            read_alg_scores(net_path, alg_names, alg_scores);
            // cerr << "Alg scores read" << endl;
            vector<string> lambda_dirs;
            string sim_net_path = base + "/" + SIMULATIONS_DIR + net_name;
            // cerr << sim_net_path << endl;
            read_directory(sim_net_path, lambda_dirs);
            for (string lambda_dir : lambda_dirs){
                cerr << lambda_dir << endl;
                string sim_net_lambda_path = sim_net_path + "/" + lambda_dir;
                compute_avg(sim_net_lambda_path);
                // cerr << "avg computed" << endl;
                vector<double> sim_scores = read_sim_scores(sim_net_lambda_path);
                // cerr << "dsd:" << sim_scores.size() << endl;
                // cerr << "sim scores read" << endl;
                vector<double> ktt = evaluate(sim_scores, alg_scores);
                // cerr << "ktt got" << endl;
                write_rankings(base, net_name, lambda_dir, alg_names, ktt);
                // cerr << "rankings written" << endl;
            }
        }
    }
    return 0;

}
